### PR TITLE
KAST-syntax unparsing of KORE terms

### PR DESCRIPTION
**Pull request by bmmoore**

### CODE REVIEW

File: kore/src/test/scala/org/kframework/kore/UnparseTest.scala [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  @Test def Tokens() {
+    Assert.assertEquals("""#token("Int","9")~>#token("String","Test")""",ToKast(intToToken(9)~>"Test"))
+  }
+
+  @Test def Variables() {
+    Assert.assertEquals("A=>B",ToKast(KRewrite(KVariable("A"),KVariable("B"))))
+  }
+
+  @Test def Precedence1() {
+    Assert.assertEquals("a()=>b()~>c()",ToKast(KRewrite('a(),'b())~>'c()))
```
Comment (dwightguth) [inline kore/src/test/scala/org/kframework/kore/UnparseTest.scala]: this doesn't look right. shouldn't ~> bind tighter than =>?

File: kore/src/test/scala/org/kframework/kore/UnparseTest.scala [hunk 2]
```diff
@@ -0,0 +1,62 @@
+package org.kframework.kore
+
+import org.junit.{Assert, Test}
+import org.kframework.attributes.Att
+import org.kframework.kore.KORE._
+
+class UnparseTest {
+  @Test def Token() {
+    Assert.assertEquals("""#token("Int","12")""",ToKast(KToken(Sort("Int"), "12")))
```
Comment (dwightguth) [inline kore/src/test/scala/org/kframework/kore/UnparseTest.scala:9]: isn't sort supposed to be second?

File: kore/src/test/scala/org/kframework/kore/UnparseTest.scala [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  @Test def EscapeLabel() {
+    Assert.assertEquals("`_+_`",ToKast(KLabel("_+_")))
+  }
+
+  @Test def WrappedKLabel() {
+    Assert.assertEquals("#klabel(foo)",ToKast(InjectedKLabel('foo, Att())))
+  }
+
+  @Test def EmptyApp() {
+    Assert.assertEquals("foo()",ToKast('foo()))
```
Comment (dwightguth) [inline kore/src/test/scala/org/kframework/kore/UnparseTest.scala]: pretty sure you need the .::KList for it to parse back.

File: kore/src/test/scala/org/kframework/kore/UnparseTest.scala [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  @Test def NestedApp() {
+    Assert.assertEquals("foo(a(),b())",ToKast('foo('a(), 'b())))
+  }
+
+  @Test def SequenceEmpty() {
+    Assert.assertEquals(".K",ToKast(ADT.KSequence(List(),Att())))
```
Comment (dwightguth) [inline kore/src/test/scala/org/kframework/kore/UnparseTest.scala:32]: .::K

File: kore/src/test/scala/org/kframework/kore/UnparseTest.scala [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  @Test def Variables() {
+    Assert.assertEquals("A=>B",ToKast(KRewrite(KVariable("A"),KVariable("B"))))
+  }
+
+  @Test def Precedence1() {
+    Assert.assertEquals("a()=>b()~>c()",ToKast(KRewrite('a(),'b())~>'c()))
+  }
+
+  @Test def Precedence2() {
+    Assert.assertEquals("a()=>``b()~>c()``",ToKast(KRewrite('a(),'b()~>'c())))
```
Comment (dwightguth) [inline kore/src/test/scala/org/kframework/kore/UnparseTest.scala]: you only need one ` here.

Comment (bmmoore) [inline kore/src/test/scala/org/kframework/kore/UnparseTest.scala]: Yes, but it's a bit complicated to figure that out. Using two should be correct.

File: kore/src/test/scala/org/kframework/kore/UnparseTest.scala (hunk 1 above)
Comment (bmmoore) [inline kore/src/test/scala/org/kframework/kore/UnparseTest.scala]: Right, I'm reversing that. It's unfortunately not described in kast.k

File: kore/src/test/scala/org/kframework/kore/UnparseTest.scala (hunk 4 above)
Comment (bmmoore) [inline kore/src/test/scala/org/kframework/kore/UnparseTest.scala:32]: .K is equivalent to .::K in kast.k - and is what's used on the wiki.

File: kore/src/test/scala/org/kframework/kore/UnparseTest.scala (hunk 2 above)
Comment (bmmoore) [inline kore/src/test/scala/org/kframework/kore/UnparseTest.scala:9]: Hmm, that is what's on the wiki, but the Kore types and the previous unparser agree on putting Sort first. I suppose I do prefer putting the sort second.

File: kore/src/test/scala/org/kframework/kore/UnparseTest.scala (hunk 3 above)
Comment (bmmoore) [inline kore/src/test/scala/org/kframework/kore/UnparseTest.scala]: .KList is what's used on the wiki, and kast.k accepts both.

**Comment by bmmoore** [comment]
Ok, I've added the explicit .KList. I think swapping the argument order on token should be left for a separate request, that will touch a lot more files if we do the KToken constructor too (and the parser/Down works by reflection so that's almost required).

**Comment by dwightguth** [comment]
lgtm

**Pull request merged**